Exclude non-content bitstream view events from Google Analytics 4

nona-luypaert opened the pull request:
## References
* Fixes #8938
* PR for dspace-7_x: https://github.com/DSpace/DSpace/pull/8949

## Description
This PR adds a config option to filter out views of non-content bitstreams (e.g. thumbnails). If the config is enabled (default), an event will only be sent to GA4 if:
- it is a view event (was already the case).
- the object of the event is a bitstream (was already the case). 
- the bitstream belongs to the 'ORIGINAL' bundle (new).

## Instructions for Reviewers

#### Changes

- `isNotBitstreamViewEvent` method in `GoogleAsyncEventListener` is now `isContentBitstream` and verifies the 3 rules described above.
- Added new `google-analytics.bundles` option in dspace.cfg to control whether to exlude/include non-content bitstreams.
- Added 2 new ITs `GoogleAsyncEventListenerIT` to test this.

#### Instructions

1. Open up your realtime view in Google Analytics
2. Visit the homepage of your repository
3. No events should be registered
4. Go to an item page with bitstreams in the 'ORIGINAL', 'LICENSE' and 'THUMBNAIL' bundles
5. Download the 3 bitstreams and verify only 1 event shows up (the one for the 'ORIGINAL' bitstream)

## Checklist

- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

nona-luypaert committed:
103837: Only send GA4 events for ORIGINAL bitstreams + ITs

nona-luypaert committed:
Merge branch 'w2p-103837_GA4-exclude-non-content-bitstream-views-7.4' into w2p-103837_GA4-exclude-non-content-bitstream-views-main

 Conflicts:
	dspace/config/dspace.cfg

nona-luypaert committed:
Merge branch 'main' into w2p-103837_GA4-exclude-non-content-bitstream-views-main

tdonohue reviewed (commented):
@nona-luypaert : Apologies it took a while to get to reviewing this.  However, my immediate feedback is *why are we making this configurable*?  

Is there any real use case to setting `google-analytics.exclude-non-content-bitstreams=false`?  It seems like, to me that no one should ever want to count *thumbnail* accesses in Google Analytics.

The only configuration I can think of that might be useful here would be to allow sites to specify which bundles are counted.  By default it should only be `ORIGINAL`.  But, if sites store content files in *other custom bundles*, then I can see a site perhaps wanting to configure this to count accesses from `ORIGINAL` and that custom bundle.  I'd consider this *optional* though.

So, overall, I think this is a great PR.  But, I don't understand why anyone would ever want to set the current config to `false`... which implies to me that we may not want to make it configurable.

nona-luypaert committed:
103837: Refactor GA config to list bundles

nona-luypaert committed:
Merge branch 'w2p-103837_GA4-exclude-non-content-bitstream-views-7.4' into w2p-103837_GA4-exclude-non-content-bitstream-views-main

nona-luypaert commented:
Hi @tdonohue , sorry it took me until now to get to this. 
I refactored the code to use a "bundle-config" (fallback: `ORIGINAL`) instead, as you suggested.
Also updated #8949.

tdonohue commented on dspace-api/src/main/java/org/dspace/google/GoogleAsyncEventListener.java:
```diff
@@ -171,9 +173,33 @@ private String getDocumentPath(HttpServletRequest request) {
+    private boolean isContentBitstream(UsageEvent usageEvent) {
+        // check if event is a VIEW event and object is a Bitstream
+        if (usageEvent.getAction() == UsageEvent.Action.VIEW
+            || usageEvent.getObject().getType() == Constants.BITSTREAM) {
```
This logic looks wrong.  It's checking if it's a `VIEW` OR a Bitstream.  I think it should be using `&&`?

tdonohue reviewed (changes requested):
@nona-luypaert : Thanks for this!  Overall, this looks good and appears to work at a basic level.  However, the `if` statement logic below appears incorrect to me... I'm concerned the logic in `isContentBitstream()` is triggering much more frequently than it should be.

Could you take a closer look at my inline comment below?

nona-luypaert committed:
103837: Fix isContentBitstream() in GoogleAsyncEventListener

nona-luypaert committed:
Merge branch 'w2p-103837_GA4-exclude-non-content-bitstream-views-7.4' into w2p-103837_GA4-exclude-non-content-bitstream-views-main

nona-luypaert commented:
Hi @tdonohue , you were indeed correct and I've fixed the method. My apologies for the mistake.

tdonohue reviewed (approved):
👍 Thanks @nona-luypaert !  This looks good to me now.  I also tested this and verified the following:
* Default settings result in only bitstream downloads from `ORIGINAL` bundle being tracked as `item` events
* Adding `google-analytics.bundles = none` results in all bitstream downloads being untracked.
* Changing to `google-analytics.bundles = ORIGINAL, THUMBNAIL` results in the same behavior as current `main`.  Both bitstream downloads and simple thumbnail views are tracked.

So, this fixes the bug & works as designed!

tdonohue commented:
Merged into `dspace-7_x` in #8949
